### PR TITLE
Add background color to tags #173

### PR DESCRIPTION
**Pull request by edmundmok**
For #173 

Preview: 
image

### CODE REVIEW

File: src/main/resources/view/PersonListCard.fxml [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
     <children>
-        <GridPane HBox.hgrow="ALWAYS" >
+        <GridPane HBox.hgrow="ALWAYS">
             <columnConstraints>
                 <ColumnConstraints hgrow="SOMETIMES" minWidth="10.0" prefWidth="150.0" />
-                <ColumnConstraints hgrow="SOMETIMES" maxWidth="100.0" minWidth="10.0" prefWidth="100.0" />
             </columnConstraints>
             <children>
-                <VBox alignment="CENTER_LEFT" maxHeight="150.0" minHeight="105.0" prefHeight="115.0" GridPane.columnIndex="0">
+                <VBox alignment="CENTER_LEFT" maxHeight="1.7976931348623157E308" minHeight="105.0" GridPane.columnIndex="0">
```
Comment (pyokagan) [inline src/main/resources/view/PersonListCard.fxml]: Huh? `1.7976931348623157E308` ? :-D

File: src/main/resources/view/PersonListCard.fxml [hunk 2]
```diff
@@ -1,18 +1,23 @@
...
+<?import javafx.scene.layout.RowConstraints?>
+<?import javafx.scene.layout.VBox?>
+
+<HBox id="cardPane" fx:id="cardPane" xmlns="http://javafx.com/javafx/8.0.60" xmlns:fx="http://javafx.com/fxml/1">
     <children>
-        <GridPane HBox.hgrow="ALWAYS" >
+        <GridPane HBox.hgrow="ALWAYS">
             <columnConstraints>
                 <ColumnConstraints hgrow="SOMETIMES" minWidth="10.0" prefWidth="150.0" />
-                <ColumnConstraints hgrow="SOMETIMES" maxWidth="100.0" minWidth="10.0" prefWidth="100.0" />
```
Comment (pyokagan) [inline src/main/resources/view/PersonListCard.fxml]: This change is fine -- multiple `ColumnConstraints` does not make any sense.

File: src/main/resources/view/PersonListCard.fxml [hunk 3]
```diff
@@ -22,21 +27,24 @@
...
                                 <HBox>
-                                    <Label fx:id="id" styleClass="cell_big_label"></Label>
-                                    <Label fx:id="name" text="\$first" styleClass="cell_big_label"/>
+                                    <Label fx:id="id" styleClass="cell_big_label" />
+                                    <Label fx:id="name" styleClass="cell_big_label" text="\$first" />
                                 </HBox>
-                                <Label fx:id="tags" styleClass="cell_small_label" text="\$tags" />
                             </children>
                         </HBox>
+                  <FlowPane fx:id="tags" />
```
Comment (pyokagan) [inline src/main/resources/view/PersonListCard.fxml]: The indentation for this should be consistent with the rest.

File: src/main/resources/view/PersonListCard.fxml [hunk 4]
```diff
@@ -22,21 +27,24 @@
...
                         <Label fx:id="phone" styleClass="cell_small_label" text="\$phone" />
                         <Label fx:id="address" styleClass="cell_small_label" text="\$address" />
                         <Label fx:id="email" styleClass="cell_small_label" text="\$email" />
                     </children>
                 </VBox>
             </children>
+         <rowConstraints>
+            <RowConstraints />
+         </rowConstraints>
```
Comment (pyokagan) [inline src/main/resources/view/PersonListCard.fxml]: Probably not needed.

File: src/main/resources/view/PersonListCard.fxml [hunk 5]
```diff
@@ -22,21 +27,24 @@
                     </padding>
 
                     <children>
-                        <HBox spacing="5" alignment="CENTER_LEFT">
+                        <HBox alignment="CENTER_LEFT" spacing="5">
```
Comment (pyokagan) [inline src/main/resources/view/PersonListCard.fxml]: Yeah, no point making a change just to swap the order of attributes :-P

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 6]
```diff
@@ -59,4 +60,7 @@ public void setNode(Node node) {
     public String getFxmlPath() {
         return FXML;
     }
+    private void initTags() {
+        person.getTags().forEach(tag -> tags.getChildren().add(new Label(tag.tagName)));
+    }
```
Comment (pyokagan) [inline src/main/java/seedu/address/ui/PersonCard.java]: OK. I do wonder what happens when [`Person::setTags()`](https://github.com/se-edu/addressbook-level4/blob/7ff3ad82caaf32e71def3d26f133c1152cad0ac6/src/main/java/seedu/address/model/person/Person.java#L68) is called though :-D I guess the UI wouldn't update in that case. Making `Person` mutable seems to be problematic, but I don't think that needs to be fixed in this PR until we get around to doing the edit command.

File: src/main/resources/view/PersonListCard.fxml (hunk 2 above)
Comment (edmundmok) [inline src/main/resources/view/PersonListCard.fxml]: There was a gridpane in the original card, which I removed since it was quite pointless (not sure about this, maybe something was planned there?). Leaving it there would lead to wasted space on the right of the card when the number of tags becomes quite large.

File: src/main/resources/view/PersonListCard.fxml (hunk 1 above)
Comment (edmundmok) [inline src/main/resources/view/PersonListCard.fxml]: MAX_VALUE for max height of the card so that it grows vertically when the number of tags is quite large.

File: src/main/resources/view/PersonListCard.fxml (hunk 5 above)
Comment (edmundmok) [inline src/main/resources/view/PersonListCard.fxml]: By the way, these changes were made using SceneBuilder... not sure why it changes these stuff, I can change it back if needed.

File: src/main/resources/view/PersonListCard.fxml (hunk 3 above)
Comment (edmundmok) [inline src/main/resources/view/PersonListCard.fxml]: I'll change this manually.

File: src/main/resources/view/PersonListCard.fxml (hunk 1 above)
Comment (pyokagan) [inline src/main/resources/view/PersonListCard.fxml]: Oh, why not just remove the `maxHeight` then?

**Comment by edmundmok** [comment]
Changed the card handle to check the tags as well by converting them into a string like how  `tagsString()` does, and checking if they are the same.

**Comment by pyokagan** [comment]
@edmundmok Sorry for the delay. I have to delay for another one day though, so sorry about that ><.

**Comment by edmundmok** [comment]
@pyokagan No problem!

File: src/test/java/guitests/guihandles/PersonCardHandle.java [hunk 7]
```diff
@@ -41,18 +41,26 @@ public String getEmail() {
         return getTextFromLabel(EMAIL_FIELD_ID);
     }
 
+    public String getTags() {
+        return getTextFromTags(node);
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: This seems a little too inflexible. What happens, for example, if I want to iterate over each tag of the person card? Maybe this should return a `List<String>` instead.

File: src/test/java/guitests/guihandles/PersonCardHandle.java [hunk 8]
```diff
@@ -41,18 +41,26 @@ public String getEmail() {
...
+    }
+
     public boolean isSamePerson(ReadOnlyPerson person) {
         return getFullName().equals(person.getName().fullName) && getPhone().equals(person.getPhone().value)
                 && getEmail().equals(person.getEmail().value)
-                && getAddress().equals(person.getAddress().value);
+                && getAddress().equals(person.getAddress().value)
+                && getTags().equals(person.tagsString());
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: I'm a bit concerned that we are comparing the *string representation* of the tags. This may make the tests too sensitive. (See "Overspecified software" in http://xunitpatterns.com/Fragile%20Test.html)

We are not concerned with the string formatting of the tags, but just whether we actually inserted each tag as a label node in the scene graph. So we should compare the tags as a list of strings or something.

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 9]
```diff
@@ -68,6 +71,29 @@ public void pressEnter() {
         guiRobot.type(KeyCode.ENTER).sleep(500);
     }
 
+    protected String getTextFromTags(Node parentNode) {
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: Since this is in the `GuiHandle` base class, *all* derived classes will get this method.

However, this method seems to require a scene graph structure that is a bit too specific to how the scene graph of the person card is implemented. This should probably be a private method of `PersonCardHandle` instead.

Also, this method requires a certain scene graph structure, so that should probably be documented in its javadoc.

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 10]
```diff
@@ -68,6 +71,29 @@ public void pressEnter() {
...
+
+        final String tagsFieldId = "#tags";
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: This should be a constant in `PersonCardHandle`.

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 11]
```diff
@@ -68,6 +71,29 @@ public void pressEnter() {
...
+            .forEach(tag -> {
+                String tagAsString = TestUtil.getTagAsString(((Label) tag).getText());
+                buffer.append(tagAsString).append(separator);
+            });
+
+        if (buffer.length() == 0) {
+            return "";
+        } else {
+            return buffer.substring(0, buffer.length() - separator.length());
+        }
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: All of these can probably be more succinctly written as (untested)
```java
        Region tagsContainer = guiRobot.from(parentNode).lookup("#tags").query();
        return tagsContainer.getChildrenUnmodifiable().stream()
            .map(node -> ((Labeled) node).getText())
            .map(TestUtil::getTagAsString)
            .collect(Collectors.joining(", "));
```

You should really consider returning a `List<String>` though.

File: src/test/java/guitests/guihandles/GuiHandle.java [hunk 12]
```diff
@@ -68,6 +71,29 @@ public void pressEnter() {
         guiRobot.type(KeyCode.ENTER).sleep(500);
     }
 
+    protected String getTextFromTags(Node parentNode) {
+
+        final String tagsFieldId = "#tags";
+        final StringBuffer buffer = new StringBuffer();
+        final String separator = ", ";
+
+        FlowPane tagsFlowPane = (FlowPane) guiRobot.from(parentNode).lookup(tagsFieldId).tryQuery().get();
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/GuiHandle.java]: Requiring a `FlowPane` is a bit too specific. The most low level class that we require is `Region`, which provides `getChildrenUnmodifiable()`.

File: src/main/resources/view/PersonListCard.fxml [hunk 13]
```diff
@@ -25,12 +30,12 @@
                         <HBox spacing="5" alignment="CENTER_LEFT">
                             <children>
                                 <HBox>
-                                    <Label fx:id="id" styleClass="cell_big_label"></Label>
-                                    <Label fx:id="name" text="\$first" styleClass="cell_big_label"/>
+                                    <Label fx:id="id" styleClass="cell_big_label" />
+                                    <Label fx:id="name" styleClass="cell_big_label" text="\$first" />
```
Comment (pyokagan) [inline src/main/resources/view/PersonListCard.fxml]: Unnecessary change: switching the order of attributes.

**Comment by damithc** [comment]
@edmundmok reminder to resume this PR

File: src/test/java/guitests/guihandles/PersonCardHandle.java (hunk 8 above)
Comment (edmundmok) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: Noted!

**Comment by edmundmok** [comment]
Sorry for the delay. Have made some changes directly into the commits and squashed them.

File: src/test/java/guitests/guihandles/PersonCardHandle.java [hunk 14]
```diff
@@ -41,18 +48,40 @@ public String getEmail() {
...
+    }
+
     public boolean isSamePerson(ReadOnlyPerson person) {
         return getFullName().equals(person.getName().fullName) && getPhone().equals(person.getPhone().value)
                 && getEmail().equals(person.getEmail().value)
-                && getAddress().equals(person.getAddress().value);
+                && getAddress().equals(person.getAddress().value)
+                && getTags().equals(person.getTags().asObservableList()
+                        .stream()
+                        .map(tag -> tag.toString())
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: `tag.toString()` adds that annoying `[` `]` brackets, which is why you need to use `getTagAsString()` above.

If you just compare `tag.tagName` directly, then you will not need to implement and use `getTagAsString()` at all.

File: src/test/java/guitests/guihandles/PersonCardHandle.java [hunk 15]
```diff
@@ -41,18 +48,40 @@ public String getEmail() {
...
+    }
+
+    private List<String> getTagsAsList(Region tagsContainer) {
+        return tagsContainer.getChildrenUnmodifiable().stream()
+                .map(node -> TestUtil.getTagAsString(((Labeled) node).getText()))
+                .collect(Collectors.toList());
+    }
+
+    private Region getTagsContainer() {
+        return (Region) guiRobot.from(node).lookup(TAGS_FIELD_ID).query();
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: No need the `(Region)` cast.

File: src/test/java/guitests/guihandles/PersonCardHandle.java [hunk 16]
```diff
@@ -41,18 +48,40 @@ public String getEmail() {
         return getTextFromLabel(EMAIL_FIELD_ID);
     }
 
+    public List<String> getTags() {
+        Region tagsContainer = getTagsContainer();
+        return getTagsAsList(tagsContainer);
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: Can just call `getTagsAsList(getTagsContainer())` directly.

File: src/test/java/guitests/guihandles/PersonCardHandle.java [hunk 17]
```diff
@@ -41,18 +48,40 @@ public String getEmail() {
...
+    }
+
+    private List<String> getTagsAsList(Region tagsContainer) {
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: Consider making this an overload of `getTags()`, e.g. `getTags(Region tagsContainer)`.

File: src/test/java/guitests/guihandles/PersonCardHandle.java [hunk 18]
```diff
@@ -41,18 +48,40 @@ public String getEmail() {
...
+
     public boolean isSamePerson(ReadOnlyPerson person) {
         return getFullName().equals(person.getName().fullName) && getPhone().equals(person.getPhone().value)
                 && getEmail().equals(person.getEmail().value)
-                && getAddress().equals(person.getAddress().value);
+                && getAddress().equals(person.getAddress().value)
+                && getTags().equals(person.getTags().asObservableList()
+                        .stream()
+                        .map(tag -> tag.toString())
+                        .collect(Collectors.toList()));
```
Comment (pyokagan) [inline src/test/java/guitests/guihandles/PersonCardHandle.java]: Maybe move all these stream processing stuff into a separate method. It can be an overload of `getTags()`, e.g. `getTags(UniqueTagList tags)`.

**Comment by pyokagan** [comment]
@edmundmok Re: code coverage. I think you can just remove `ReadOnlyPerson::tagsString()`.

**Comment by damithc** [comment]
Good work @edmundmok 👍

**Comment by edmundmok** [comment]
Thanks @pyokagan for the helpful comments and guidance! I've learnt a lot from this.
